pkg/models/vulnerability.go

```diff
@@ -167,3 +174,130 @@ func (v Vulnerability) MarshalYAML() (interface{}, error) {
 
 	return raw, nil
 }
+
+func (e Event) version() string {
+	if e.Introduced != "" {
+		return e.Introduced
+	}
+
+	if e.Fixed != "" {
+		return e.Fixed
+	}
+
+	if e.Limit != "" {
+		return e.Limit
+	}
+
+	if e.LastAffected != "" {
+		return e.LastAffected
+	}
+
+	return ""
+}
+
+func (ar Range) containsVersion(pkg lockfile.PackageDetails) bool {
+	if ar.Type != RangeEcosystem && ar.Type != RangeSemVer {
+		return false
+	}
+	// todo: we should probably warn here
+	if len(ar.Events) == 0 {
+		return false
+	}
+
+	vp := semantic.MustParse(pkg.Version, pkg.CompareAs)
+
+	sort.Slice(ar.Events, func(i, j int) bool {
+		return semantic.MustParse(ar.Events[i].version(), pkg.CompareAs).CompareStr(ar.Events[j].version()) < 0
+	})
+
+	var affected bool
+	for _, e := range ar.Events {
+		if affected {
+			if e.Fixed != "" {
+				affected = vp.CompareStr(e.Fixed) < 0
+			} else if e.LastAffected != "" {
+				affected = e.LastAffected == pkg.Version || vp.CompareStr(e.LastAffected) <= 0
+			}
+		} else if e.Introduced != "" {
+			affected = e.Introduced == "0" || vp.CompareStr(e.Introduced) >= 0
+		}
+	}
+
+	return affected
+}
+
+// affectsVersion checks if the given version is within the range
+// specified by the events of any "Ecosystem" or "Semver" type ranges
+func affectsVersion(a []Range, pkg lockfile.PackageDetails) bool {
+	for _, r := range a {
+		if r.Type != RangeEcosystem && r.Type != RangeSemVer {
+			return false
+		}
+		if r.containsVersion(pkg) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (v Vulnerability) isAliasOfID(id string) bool {
+	for _, alias := range v.Aliases {
+		if alias == id {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (v Vulnerability) isAliasOf(vulnerability Vulnerability) bool {
+	for _, alias := range vulnerability.Aliases {
+		if v.ID == alias || v.isAliasOfID(alias) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (v Vulnerability) AffectsEcosystem(ecosystem lockfile.Ecosystem) bool {
+	for _, affected := range v.Affected {
+		if string(affected.Package.Ecosystem) == string(ecosystem) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (v Vulnerability) IsAffected(pkg lockfile.PackageDetails) bool {
+	for _, affected := range v.Affected {
+		if string(affected.Package.Ecosystem) == string(pkg.Ecosystem) &&
```

Comment:
nit: this function is getting pretty indented. consider inverting this condition and continuing instead.

Reply:
I'm not sure it'd help matters much because the conditions would then be inverted which is typically considered hard to read

Reply:
Fair enough. I typically prefer early exiting to avoid having too indented code which ends up being difficult to read. Another alternative is to just factor our the indented bits into another function.

No strong opinions here so happy to leave as is.